src/core/classes/prism.ts

```
@@ -0,0 +1,132 @@
+import globalDefaults, { type PrismConfig } from './config';
+import { Hooks } from './hooks';
+import { Registry } from '../registry';
+import { highlightAll, type HighlightAllOptions } from '../highlight-all';
+import { highlightElement, type HighlightElementOptions } from '../highlight-element';
+import { highlight, type HighlightOptions } from '../highlight';
+import { tokenize } from '../tokenize';
+import { nextTick, documentReady, allSettled } from '../../util';
+import PluginRegistry from './plugin-registry';
+import LanguageRegistry from './language-registry';
+import type { Grammar, LanguageProto, PluginProto } from '../../types';
+import type { TokenStream } from '../token';
+
+/**
+ * Prism class, to create Prism instances with different settings.
+ * In most use cases, you just need the pre-existing Prism instance, see {@link prism}.
+ */
+export default class Prism {
+	hooks = new Hooks();
+
+	// TODO remove this and make sure the functionality is covered by language and plugin registries
+	components = new Registry(this);
+	languageRegistry: LanguageRegistry;
+	pluginRegistry: PluginRegistry;
+
+	config: PrismConfig;
+
+	languagesReady: Promise<unknown>;
+	waitFor: Promise<unknown>[] = [nextTick()];
+	ready: Promise<unknown> = allSettled(this.waitFor);
+
+	constructor (config: PrismConfig = {}) {
+		this.config = Object.assign({}, globalDefaults, config);
+		this.config.errorHandler ??= this.config.silent ? () => undefined : console.error;
+
+		const reportError: (reason: any) => PromiseLike<never> = this.config.errorHandler;
+
+		this.languageRegistry = new LanguageRegistry(this.config.languagePath ?? './languages/');
+		this.pluginRegistry = new PluginRegistry(this.config.pluginPath ?? './plugins/');
+
+		// Preload languages
+		if (this.config.languages?.length > 0) {
```

```suggestion
		// @ts-expect-error: TypeScript does not narrow the type of the languages property when using optional chaining in the condition.
		if (this.config.languages?.length > 0) {
```

TS drives me crazy sometimes. 🤪 We can either go the proposed way and ignore this error, since our code clearly shows our intention. Or we can help TS to narrow down types like so:

```ts
const languages = this.config.languages;
if (languages && languages.length > 0) {
	this.languageRegistry.loadAll(languages);
}
```

In that case, we can ignore the following suggestion (about adding `!`).

---

Let's do that (`languages && languages.length > 0`)